Feat: Add functions to access insecticide resistance phenotype data

mohamed-laarej opened the pull request:
**Feat: Add functions to access insecticide resistance phenotype data**

**Resolves:** #789

**Description:**

This PR introduces new functionality to access insecticide resistance phenotype data, as requested in issue #789 for GSoC 2025. The primary goal is to provide users with easy access to bioassay results (e.g., alive/dead status after insecticide exposure) alongside sample metadata and optionally linked genetic data.

**Changes Implemented:**

1.  **New Mixin Class:** Created `malariagen_data.anoph.phenotypes.PhenotypeDataMixin` located in the new file `malariagen_data/anoph/phenotypes.py`. This mixin encapsulates all logic related to fetching, processing, and merging phenotype data.
2.  **Integration with Ag3:** The main `Ag3` class (in `malariagen_data/ag3.py`) now inherits from `PhenotypeDataMixin`. This seamlessly integrates the new phenotype methods, making them directly available on `Ag3` instances (e.g., `ag3.phenotype_data(...)`).

**New Functionality Provided by `PhenotypeDataMixin`:**

The following public methods are now available via `Ag3` objects:

*   `phenotype_data(...)`: Loads insecticide bioassay data for specified sample sets from GCS. It merges this with sample metadata and returns a pandas DataFrame. Users can filter by `sample_sets`, `sample_query`, `insecticide`, `dose`, `phenotype`, and apply cohort size constraints (`cohort_size`, `min_cohort_size`, `max_cohort_size`).
*   `phenotype_binary(...)`: Similar to `phenotype_data`, but processes the results to return a pandas Series containing binary phenotype outcomes (1 = alive/resistant, 0 = dead/susceptible), indexed by `sample_id`. This is useful for downstream analyses requiring binary resistance status.
*   `phenotypes(...)`: Provides a comprehensive view by loading phenotype data and optionally merging it with genetic variant data (`snp_calls` or `haplotypes`) for a specified `region`. It returns an xarray Dataset containing both phenotype and genetic information for common samples. The function handles cases where genetic data is unavailable or does not overlap with the phenotype samples for the given query.
*   `phenotype_sample_sets()`: A utility method that returns a list of sample sets for which phenotype data files were found in GCS.

Several internal helper methods (`_load_phenotype_data`, `_merge_phenotype_with_metadata`, `_create_phenotype_binary_series`, `_create_phenotype_dataset`, etc.) support these public functions.

**Testing:**

Local testing using a script confirmed that `phenotype_data` and `phenotype_binary` load and process data correctly. The `phenotypes` function was also tested; the logic for merging phenotype data with genetic data appears sound and correctly handles cases where genetic data is unavailable or does not overlap with the available phenotype samples. However, fully verifying the merge output with actual overlapping data was challenging due to data availability limitations in the local test environment using the currently accessible sample sets.

Formal unit tests covering the new functions will be added as per project guidelines and mentor recommendations following this initial review.

**Next Steps:**

*   Requesting review of the implementation approach and code.
*   Plan to add formal unit tests covering the new functions as per project guidelines after initial feedback.

mohamed-laarej commented:
Hi @jonbrenas ,

Thank you very much for taking the time to review this PR and for your valuable feedback!
I've pushed new commits addressing the points you raised.
Please let me know if these changes look good or if there's anything else. Thanks again!

jonbrenas commented on malariagen_data/anoph/phenotype_params.py:
```diff
@@ -0,0 +1,13 @@
+from typing import TypeAlias, Optional, Union, List
+
+# Type alias for insecticide parameter
+insecticide: TypeAlias = Optional[Union[str, List[str]]]
```
These should be 'Annotated` to help with the docs.

jonbrenas commented on malariagen_data/anoph/phenotypes.py:
```diff
@@ -0,0 +1,505 @@
+    Provides methods for accessing insecticide resistance phenotypic data.
+    Inherited by AnophelesDataResource subclasses (e.g., Ag3).
+    """
+
```
I think you are missing the constructor here.

jonbrenas reviewed (changes requested):
Thanks @mohamed-laarej. 

In addition to the tests that were mentioned during our call, I added a few comments here.

jonbrenas commented:
Thanks @mohamed-laarej. 

I ran a quick debug of your code and the reason why you don't get the `snp_calls` in `phenotypes` is that, in `_create_phenotype_dataset`, you use `variant_data.coords["samples"].values` as the list of sample ids from `variant_data` instead of `variant_data.coords["sample_id"].values`. I haven't looked at why you can't find the haplotypes yet but I am sure you will have it figured out before I get to it.

mohamed-laarej commented:
Thanks @jonbrenas for the detailed feedback! Really helpful stuff. Let me go through each point:
## Functions as Global Variables
Ah yeah, I can see why that looked weird! I ended up doing it that way because I was getting a bunch of MyPy errors when I first tried the mixin approach:
```console
malariagen_data\anoph\phenotypes.py:31: error: "AnophelesPhenotypeData" has no attribute "_url" [attr-defined]
malariagen_data\anoph\phenotypes.py:377: error: "AnophelesPhenotypeData" has no attribute "sample_metadata" [attr-defined]
```
To fix this, I added type annotations and made them constructor params when you suggested adding a constructor. But after reviewing how the other mixins work, I refactored it to use cooperative inheritance like this:
```python
class AnophelesPhenotypeData:
    # Type annotations for MyPy
    _url: str
    _fs: fsspec.AbstractFileSystem
    sample_metadata: Callable[..., pd.DataFrame]
    # etc...
    
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
```
I think it's much cleaner! Now it just calls `self.sample_metadata()` directly like it should.
Could you please confirm if this is the right approach? Or is there anything else I should consider here to make it more idiomatic or robust?

## Sample Query Filtering
Completely agree on this one! Consolidating everything into `sample_query` will definitely make the API cleaner and more consistent with how the rest of the codebase works.
## Clearer Data Type Selection
This is a great point, the current logic is definitely confusing. I'd love your thoughts on which approach would work best:
### Option A - Separate functions:

```python
def phenotypes_with_snps(self, region, sample_query=None, ...):
def phenotypes_with_haplotypes(self, region, sample_query=None, analysis="arab", ...):
def phenotypes_with_cnvs(self, region, sample_query=None, ...):
def phenotypes_only(self, sample_query=None, ...):
```
### Option B - Single function with parameter:
```python
def phenotypes(self, region=None, genetic_data="snps", sample_query=None, ...):
    # where genetic_data can be "snps", "haplotypes", "cnvs", or None
```
I'm personally leaning toward Option A since it makes the intent really clear and avoids any ambiguity, but I'd definitely defer to your judgment on what would be most intuitive for users.

## Next Steps
I'll get started on:
Cleaning up the inheritance structure based on your guidance
Moving all filtering logic to ` sample_query`
Implementing whichever genetic data selection approach you prefer
Putting together a demo notebook to show how everything works

Thanks again for taking the time to provide such detailed feedback, it's really helpful for understanding the codebase patterns better!

jonbrenas commented on malariagen_data/anopheles.py:
```diff
@@ -52,7 +52,7 @@
     plotly_discrete_legend,
 )
-
+from .anoph.phenotypes import AnophelesPhenotypeData
```
Shift to before the functions import to be consistent.

mohamed-laarej replied:
Thanks @jonbrenas ,
I believe that part was added due to a `mypy` issue or possibly a test failure related to `site_mask` values when using ["`gamb_colu_arab`", "`gamb_colu`", "`arab`"] across all fixture types. I don't recall the exact error message, but I think `mypy` (or the tests) was unhappy when some simulators were paired with incompatible or undefined site masks. That’s why I added the` isinstance` check to restrict values to valid options per simulator.

Happy to revisit this and simplify it if needed.

jonbrenas reviewed (changes requested):
Great work, @mohamed-laarej. In `phenotype_data`, `sample_query_options` isn't used properly: it is passed to `sample_metadata` without a query (i.e., it does nothing) and not used when the query is actually being evaluated.

jonbrenas reviewed (approved):
LGTM

leehart commented:
I've always assumed that the blank line between imports at the top of `ag3.py` is used to visually separate external from internal imports. For comparison, there are also blank lines between the imports in `af1.py`, which I took to be visual separations for standard library modules (e.g. `sys`) and external modules (e.g. `plotly.express`) and internal modules (e.g. `.anopheles`).

This might be a style choice and does not need to be changed in this PR. Although the style does not seem to be applied consistently everywhere (e.g. `anoph/aim_data.py`), perhaps because it was never written as policy.

If we decide to police "don't have lines between imports", then that would require a lot of changes. Normally this wouldn't even be noticeable or noteworthy, but there is no other change in `ag3.py` in this PR.

mohamed-laarej commented:
Hi @jonbrenas, I just wanted to give you a quick heads-up
While reviewing the PR after approval (thank you again!), I realized that the `phenotype_binary` method had been inadvertently removed during the recent refactoring around `sample_query` and `phenotype_data`, and I unfortunately overlooked re-adding it.
Since this method was part of the original functionality and provides a convenient way to access binary phenotype outcomes, I’ve restored it in line with the new design principles, it now uses `sample_query` for filtering. I’ve also added integration tests and updated the demo notebook with relevant examples.
Apologies for the late addition to an already-approved PR. If you’d prefer this be handled in a separate PR instead, I’m happy to make that change just let me know.
Thanks again for your time and understanding!

jonbrenas commented:
That's good thanks @mohamed-laarej !
